File: core/src/main/java/com/ustadmobile/core/controller/ScormPackagePresenter.java
package com.ustadmobile.core.controller;

import com.ustadmobile.core.impl.UmCallback;
import com.ustadmobile.core.impl.UstadMobileSystemImpl;
import com.ustadmobile.core.impl.http.UmHttpCall;
import com.ustadmobile.core.impl.http.UmHttpRequest;
import com.ustadmobile.core.impl.http.UmHttpResponse;
import com.ustadmobile.core.impl.http.UmHttpResponseCallback;
import com.ustadmobile.core.contentformats.scorm.ScormManifest;
import com.ustadmobile.core.util.UMFileUtil;
import com.ustadmobile.core.view.ScormPackageView;
import com.ustadmobile.core.view.UstadViewWithNotifications;

import org.xmlpull.v1.XmlPullParserException;

import java.io.IOException;
import java.util.Hashtable;

/**
 *
 * Present a SCORM package. Scorm tech specs as per :
 *  https://scorm.com/scorm-explained/technical-scorm/
 *
 * Created by mike on 1/6/18.
 */
public class ScormPackagePresenter extends UstadBaseController {

    private ScormManifest scormManifest;

    private ScormPackageView scormPackageView;

    private String mountedPath;

    private UmCallback<String> zipMountedCallback = new UmCallback<String>() {
        @Override
        public void onSuccess(String result) {
            mountedPath = result;
            UstadMobileSystemImpl.getInstance().makeRequestAsync(new UmHttpRequest(
                            getContext(),
                            UMFileUtil.joinPaths(new String[]{mountedPath, "imsmanifest.xml"})),
                    manifestLoadedCallback);
        }

        @Override
        public void onFailure(Throwable exception) {
            scormPackageView.showNotification("ERROR: failed to open package file",
                    UstadViewWithNotifications.LENGTH_LONG);
        }
    };

    private UmHttpResponseCallback manifestLoadedCallback = new UmHttpResponseCallback(){
        @Override
        public void onComplete(UmHttpCall call, UmHttpResponse response) {
            scormManifest = new ScormManifest();
            try {
                scormManifest.loadFromInputStream(response.getResponseAsStream());
                ScormManifest.Organization defaultOrg = scormManifest.getDefaultOrganization();
                final ScormManifest.Resource startRes = scormManifest.getResourceByIdentifier(
                        defaultOrg.getItems().get(0).getIdentifierRef());
                scormPackageView.runOnUiThread(new Runnable() {
                    @Override
                    public void run() {
                        scormPackageView.setTitle(scormManifest.getDefaultOrganization().getTitle());
                        scormPackageView.loadUrl(UMFileUtil.joinPaths(new String[]{mountedPath,
                                startRes.getHref()}));
                    }
                });
            }catch(IOException e) {
                e.printStackTrace();
            }catch(XmlPullParserException x) {
                x.printStackTrace();
            }
        }

        @Override
        public void onFailure(UmHttpCall call, IOException exception) {

        }
    };

    public ScormPackagePresenter(Object context, ScormPackageView view) {
        super(context);
        this.scormPackageView = view;
    }

    public void onCreate(Hashtable args) {
        scormPackageView.mountZip((String)args.get(EpubContentPresenter.ARG_CONTAINERURI),
                zipMountedCallback);
    }

}

File: core/src/main/java/com/ustadmobile/core/controller/ScormPackagePresenter.kt
package com.ustadmobile.core.controller

import com.ustadmobile.core.contentformats.scorm.ScormManifest
import com.ustadmobile.core.impl.UmCallback
import com.ustadmobile.core.impl.UstadMobileSystemImpl
import com.ustadmobile.core.impl.http.UmHttpCall
import com.ustadmobile.core.impl.http.UmHttpRequest
import com.ustadmobile.core.impl.http.UmHttpResponse
import com.ustadmobile.core.impl.http.UmHttpResponseCallback
import com.ustadmobile.core.util.UMFileUtil
import com.ustadmobile.core.view.ScormPackageView
import com.ustadmobile.core.view.UstadView
import com.ustadmobile.core.view.UstadViewWithNotifications

import org.xmlpull.v1.XmlPullParserException

import java.io.IOException
import java.util.HashMap

/**
 *
 * Present a SCORM package. Scorm tech specs as per :
 * https://scorm.com/scorm-explained/technical-scorm/
 *
 * Created by mike on 1/6/18.
 */
class ScormPackagePresenter(context: Any, arguments: Map<String, String?>, view: ScormPackageView)
    : UstadBaseController<ScormPackageView>(context, arguments, view) {

    private var scormManifest: ScormManifest? = null

    private var mountedPath: String? = null

    private val zipMountedCallback = object : UmCallback<String> {
        override fun onSuccess(result: String?) {
            mountedPath = result
            UstadMobileSystemImpl.instance.makeRequestAsync(UmHttpRequest(
                    context,
                    UMFileUtil.joinPaths(mountedPath!!, "imsmanifest.xml")),
                    manifestLoadedCallback)
        }

        override fun onFailure(exception: Throwable?) {
            view.showNotification("ERROR: failed to open package file",
                    UstadViewWithNotifications.LENGTH_LONG)
        }
    }

    private val manifestLoadedCallback = object : UmHttpResponseCallback {
        override fun onComplete(call: UmHttpCall, response: UmHttpResponse) {
            scormManifest = ScormManifest()
            try {
                scormManifest!!.loadFromInputStream(response.responseAsStream!!)
                val defaultOrg = scormManifest!!.defaultOrganization
                val startRes = scormManifest!!.getResourceByIdentifier(
                        defaultOrg.items[0].identifierRef!!)
                view.runOnUiThread(Runnable  {
                    view.setTitle(scormManifest!!.defaultOrganization.title!!)
                    view.loadUrl(UMFileUtil.joinPaths(mountedPath!!,
                            startRes.href!!))
                })
            } catch (e: IOException) {
                e.printStackTrace()
            } catch (x: XmlPullParserException) {
                x.printStackTrace()
            }

        }

        override fun onFailure(call: UmHttpCall, exception: IOException) {

        }
    }

    fun onCreate(args: HashMap<String, String>) {
        view.mountZip(args[UstadView.ARG_CONTAINER_UID]!!,
                zipMountedCallback)
    }

}
